### ISSUE TITLE
[tf2] Callbacks used in self-defined training

### ISSUE THREAD

**Issue by Yablon** [tensorflow/tensorflow#33844]
<em>Please make sure that this is a feature request. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:feature_template</em>

**System information**
- TensorFlow version (you are using): 2.0.0
- Are you willing to contribute it (Yes/No): No

**Describe the feature and the current behavior/state.**
For now,  callbacks can only be used in tf.keras.Model.fit method.
If one writes a self-defined loop, he has to write his own codes to save model, to show logs on tensorboard, and maybe many other things. 
I think if the callbacks are here, why can't we offer some methods to make it usable in self-defined loops ? Actually, self-defined loops also has concepts like batch_begin, epoch_end, etc.
This will make self-defined loops simple and easy to write. One can focus on many other things he has to do.
Thank you !

**Will this change the current api? How?**

I think there is no need to change the current api.

**Who will benefit with this feature?**

I think the people train with self-defined loops wil benefit.

**Any Other info.**

**Comment by pandrey-fr** [comment]
@Yablon Did you find a solution that had you close this issue? If so, I would be interested in knowing what your solution is, as I am currently trying to generalize the passing of callbacks to a keras-like custom training loop.

**Comment by faroit** [comment]
@Yablon I agree that this should be reopened. With full distributed training support, TF2 only recommends two ways for training which is `.fit` and custom (gradient tape) training loops. 

Also I think the callbacks can easily be rewritten so that they are available from custom training loops. E.g. think of using a learning rate scheduler

```python
sched = ReduceLROnPlateau()

**Comment by Yablon** [comment]
@pandrey-fr @faroit Sorry for a so late reply. I thought the request made me feel myself stupid and I closed it. I thought people skilled at writing codes might write the callback things easily. And I thought maybe I could turn to tf2 later...
At last, I hope callbacks can be used in self-defined training loops in the future. If we can't, maybe there is a better choice.
Thank you! Hope you be safe and sound.

**Comment by faroit** [comment]
@pandrey-fr @Yablon maybe this might not be necessary anymore with tf2.2 where users can add their custom [train_step to the fit function](https://www.tensorflow.org/versions/r2.2/api_docs/python/tf/keras/Model#fit)...

**Comment by pandrey-fr** [comment]
@faroit I agree, TF 2.2 solves this issue - well, at least, it provides a general solution to it, which is really great!

**Comment by Yablon** [comment]
Thank you, I learned a lot ! I should think about turning to tf2 now !
